src/api-service/__app__/agent_events/__init__.py

```diff
@@ -82,6 +92,20 @@ def on_state_update(
                             state=NodeTaskState.setting_up,
                         )
                         node_task.save()
+            elif state == NodeState.done:
+                # Model-validated.
+                #
+                # This field will be required in the future.
+                # For now, it is optional for back compat.
+                done_data = cast(Optional[NodeDoneEventData], state_update.data)
+
+                if done_data:
```

Comment:
This check is what lets us detect server-side that a node went to `done` because the setup script failed. Legacy `done` events won't have a `data` payload, so we can't tell why they transitioned. But they'll also never have transitioned due to a failed setup script (that's new to this PR), so it doesn't matter.